### DISCUSSION TITLE
Why can't I follow other ActivityPub sites if their host resovles to a private IP address?

### DISCUSSION THREAD

**Discussion by ncot-tech** [mastodon/mastodon#24349] [Q&A]
So I run a Mastodon instance on a Raspberry Pi. On another Pi I have Wordpress with the ActivityPub plugin. And on yet another Pi I am running PixelFed. They're all behind an NGINX reverse proxy, with one public IP address. I run a local DNS that resolves each domain to the relevant private IP address of the Pi running that site so that hosts inside my network can use the same public domains.

Trying to follow the Wordpress or PixelFed accounts from Mastodon causes a "422 HostValidationError". After much searching I found change #6410 which causes this. Seems you can't follow an instance if its domain entry resolves to a private IP address.

Why? Is there some security issue with allowing an instance to follow other locally hosted instances or something?

If I stop my internal DNS and force everything to resolve to my external IP address, it works fine.

**Comment by SoniEx2** [comment]
#24432 would help. but this exists to prevent/mitigate SSRF in misconfigured hosts.

we think you can disable it if you know what you're doing tho.

**Comment by nightpool** [comment]
> Why? Is there some security issue with allowing an instance to follow other locally hosted instances or something

Yes, this is a know issue called Server-side request forgery. For example, if you had an internally hosted elasticsearch server (common to Mastodon deployments), anybody on the internet could use the federation tools to make unauthenticated requests to your Elasticsearch server that look like they're coming from your Mastodon server.

Additionally, in the setup you're describing, users could use the follow requests to make requests that bypass any security proxy, DDOS protection or caching you have on your Pixelfed instance. For example, if you're using a WAF to protect against a security vulnerability (a common tactic when patches aren't yet available or hard to apply, for example with the Log4j vulnerability in the Java ecosystem), then allowing users to make requests from your Mastodon server to your Pixelfed server would bypass that protection.

Instead of having your external domain name change based on what network you're connected to, the best practice would be to have a separate internal hostname for any internal requests you need to make, and only use that hostname in contexts where internal requests are appropriate (e.g. from a reverse proxy to a endpoint application), not reusing the same "external" hostname but resolving it to a different IP address. Conceptually, your hostname should never "mean" two different things depending on which DNS server you connect to. In your setup, your hostname *either* connects you to nginx (a reverse proxy) *or* connects you to the application server directly. Not only can this can lead to confusing and hard to debug issues when you're not sure which host you're actually connecting to, it means you have to "implement everything twice" with respect to your nginx configuration, since there's no guarantee that nginx will be included in the path of the user's request

**Reply by SoniEx2** [reply]
if you're using an SNI proxy then none of that really applies.
